src/extensions/Wyam.YouTube/YouTube.cs

```diff
@@ -0,0 +1,113 @@
+    /// </remarks>
+    /// <category>Metadata</category>
+    public class YouTube : IModule, IAsNewDocuments
+    {
+        private readonly BaseClientService.Initializer _initializer;
+        private readonly Dictionary<string, Func<IDocument, IExecutionContext, YouTubeService, object>> _requests
+            = new Dictionary<string, Func<IDocument, IExecutionContext, YouTubeService, object>>();
+
+        /// <summary>
+        /// Creates a connection to the YouTube API with authenticated access.
+        /// </summary>
+        /// <param name="apiKey">The apikey to use.</param>
+        public YouTube(string apiKey)
+        {
+            _initializer = new BaseClientService.Initializer()
+            {
+                ApplicationName = "Wyam",
+                ApiKey = apiKey
+            };
+        }
+
+        /// <summary>
+        /// Submits a request to the YouTube client. This allows you to incorporate data from the execution context in your request.
+        /// </summary>
+        /// <param name="key">The metadata key in which to store the return value of the request function.</param>
+        /// <param name="request">A function with the request to make.</param>
+        /// <returns>The current module instance.</returns>
+        public YouTube WithRequest(string key, Func<IExecutionContext, YouTubeService, object> request)
+        {
+            if (string.IsNullOrEmpty(key))
+            {
+                throw new ArgumentException("Argument is null or empty", nameof(key));
+            }
+            if (request == null)
+            {
+                throw new ArgumentNullException(nameof(request));
+            }
+
+            _requests[key] = (doc, ctx, github) => request(ctx, github);
+            return this;
+        }
+
+        /// <summary>
+        /// Submits a request to the YouTube client. This allows you to incorporate data from the execution context and current document in your request.
+        /// </summary>
+        /// <param name="key">The metadata key in which to store the return value of the request function.</param>
+        /// <param name="request">A function with the request to make.</param>
+        /// <returns>The current module instance.</returns>
+        public YouTube WithRequest(string key, Func<IDocument, IExecutionContext, YouTubeService, object> request)
+        {
+            if (string.IsNullOrEmpty(key))
+            {
+                throw new ArgumentException("Argument is null or empty", nameof(key));
+            }
+            if (request == null)
+            {
+                throw new ArgumentNullException(nameof(request));
+            }
+
+            _requests[key] = request;
+            return this;
+        }
+
+        /// <inheritdoc />
+        public IEnumerable<IDocument> Execute(IReadOnlyList<IDocument> inputs, IExecutionContext context)
+        {
+            // YouTubeService is IDisposable.... how do you want this handled?
```

Comment:
Hmm, that's a good question. Normally I'd say just create and dispose the service instance within the execute method since the documents typically only carry the result metadata with them. But in this case since you're modeling it after the GitHub module and exposing the service as a leaky abstraction, the question of how to create and dispose the service is trickier. I might need to make a callback or event or something in the execution context to let modules clean up service classes like this when the full execution is done (documents and their metadata could be used any time up until then). Let me think on it and get back to you - for now just don't worry about disposal and we'll circle back.